cmd/oras/internal/option/format.go

```diff
@@ -15,21 +15,120 @@ limitations under the License.
+		Name:  "json",
+		Usage: "Print in JSON format",
+	}
+	FormatTypeGoTemplate = &FormatType{
+		Name:  "go-template",
+		Usage: "Print in JSON format",
+	}
+	FormatTypeTable = &FormatType{
+		Name:  "table",
+		Usage: "Get direct referrers and output in table format",
+	}
+	FormatTypeTree = &FormatType{
+		Name:  "tree",
+		Usage: "Get referrers recursively and print in tree format",
+	}
+)
+
+// Format contains input and parsed options for formatted output flags.
 type Format struct {
-	Template string
+	FormatFlag   string
+	Type         string
+	Template     string
+	AllowedTypes []*FormatType
 }
 
 // ApplyFlag implements FlagProvider.ApplyFlag.
 func (opts *Format) ApplyFlags(fs *pflag.FlagSet) {
-	const name = "format"
-	if fs.Lookup(name) != nil {
-		// allow command to overwrite the flag
-		return
-	}
-	fs.StringVar(&opts.Template, name, "", `[Experimental] Format output using a custom template:
-'json':       Print in JSON format
-'$TEMPLATE':  Print output using the given Go template.`)
+	var buf bytes.Buffer
+	_, _ = buf.WriteString("[Experimental] Format output using a custom template:")
+	w := tabwriter.NewWriter(&buf, 0, 0, 2, ' ', 0)
+	for _, t := range opts.AllowedTypes {
+		_, _ = fmt.Fprintf(w, "\n'%s':\t%s", t.Name, t.Usage)
+	}
+	w.Flush()
+	// apply flags
+	fs.StringVar(&opts.FormatFlag, "format", opts.FormatFlag, buf.String())
+	fs.StringVar(&opts.Template, "template", "", "[Experimental] Template string used to format output")
+}
+
+// Parse parses the input format flag.
+func (opts *Format) Parse(_ *cobra.Command) error {
+	if err := opts.parseFlag(); err != nil {
+		return err
+	}
+
+	if opts.Type == "" {
+		// flag not specified
+		return nil
+	}
+
+	if opts.Type == FormatTypeGoTemplate.Name && opts.Template == "" {
+		return &oerrors.Error{
+			Err:            fmt.Errorf("%q format specified but no template given", opts.Type),
+			Recommendation: fmt.Sprintf("use `--format %s=TEMPLATE` or `--format %s --template TEMPLATE` to specify the template", opts.Type, opts.Type),
+		}
+	}
+
+	var optionalTypes []string
+	for _, t := range opts.AllowedTypes {
+		if opts.Type == t.Name {
+			// type validation passed
+			return nil
+		}
+		optionalTypes = append(optionalTypes, t.Name)
+	}
+	return &oerrors.Error{
+		Err:            fmt.Errorf("invalid format type: %q", opts.Type),
+		Recommendation: fmt.Sprintf("supported types: %s", strings.Join(optionalTypes, ", ")),
+	}
+}
+
+func (opts *Format) parseFlag() error {
+	opts.Type = opts.FormatFlag
+	if opts.Template != "" {
+		// template explicitly set
+		if opts.Type != FormatTypeGoTemplate.Name {
+			return fmt.Errorf("--template must be used with --format %s", FormatTypeGoTemplate.Name)
+		}
+		return nil
+	}
+
+	goTemplatePrefix := FormatTypeGoTemplate.Name + "="
+	if strings.HasPrefix(opts.FormatFlag, goTemplatePrefix) {
+		// add parameter to template
+		opts.Type = FormatTypeGoTemplate.Name
+		opts.Template = opts.FormatFlag[len(goTemplatePrefix):]
+	}
+	return nil
 }
```

Comment:
nit: this is accepted for now. It requires refactoring if we want to support more type-value pairs in the future like `--format go-template-file=<path>`.

Reply:
Added todo comments